arroyo/processing/run_task.py

```diff
@@ -0,0 +1,77 @@
+import logging
+import time
+from collections import deque
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import Callable, Deque, Optional, Tuple
+
+from arroyo.processing.strategies.abstract import MessageRejected, ProcessingStrategy
+from arroyo.types import Commit, Message, Position, TPayload
+
+logger = logging.getLogger(__name__)
+
+
+class RunTask(ProcessingStrategy[TPayload]):
+    """
+    Runs a task in a thread. Commits offsets once futures are done.
+    """
+
+    def __init__(
+        self,
+        processing_function: Callable[[Message[TPayload]], None],
+        concurrency: int,
+        commit: Commit,
+    ) -> None:
+        self.__executor = ThreadPoolExecutor(max_workers=concurrency)
+        self.__function = processing_function
+        self.__futures: Deque[Tuple[Message[TPayload], Future[None]]] = deque()
+        self.__max_pending_futures = concurrency * 2
+        self.__commit = commit
+        self.__closed = False
+
+    def submit(self, message: Message[TPayload]) -> None:
+        assert not self.__closed
+        # The list of pending futures is too long, tell the stream processor to slow down
+        if len(self.__futures) > self.__max_pending_futures:
+            raise MessageRejected
+
+        self.__futures.append(
+            (message, self.__executor.submit(self.__function, message))
+        )
+
+    def poll(self) -> None:
+        # Remove completed futures in order
+        while self.__futures and self.__futures[0][1].done():
+            message, _ = self.__futures.popleft()
```

Comment:
I think we should elaborate a bit more this scenario.
Here we just check for completed futures and commit. We do not cover the case where the task fails and do not support a return value.
I would say managing the error case should be mandatory.
About the return value, I think that will be a need for most consumers as we generally either process a message and produce on a topic or batch processed message and write somewhere. The case where we just fork a task in a separate thread does not seem very common to me.

Also keep in mind that running a task in a separate thread is also not very common. We would not do that for high throughput consumers, we would have to batch messages first otherwise we would impact negatively our throughput with scheduling activity.

Reply:
Added the exception handling and some more documentation.

I intentionally avoided supporting a giving the user back a return value or chaining another step after this one. It's not meant for those use cases. It's only for running a simple IO-bound task and immediately committing offsets afterwards.

There are use cases for this already. I needed it here https://github.com/getsentry/sentry/pull/39502 for the post process forwarder but unfortunately it was not ready yet. I also saw some use cases where the batch processing strategy was being used where batching was not needed and a simple RunTask would have been better.

Yes - there are some scenarios where it's not appropriate, but it's not supposed to be an allen key that can fit every situation.